openshift/release/fetch-pipeline.sh

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env bash
+#
+# Detect which version of pipeline should be installed
+# First it tries nightly
+# If that doesn't work it tries previous releases (until the MAX_SHIFT variable)
+# If not it exit 1
+# It can take the argument --only-stable-release to not do nightly but only detect the pipeline version
+
+# set max shift to 0, so that when a version is explicitly specified that version is fetched
+# modify this in future if a workflow based on latest version and recent (shifted) versions is needed
+MAX_SHIFT=1
+NIGHTLY_RELEASE="https://raw.githubusercontent.com/openshift/tektoncd-pipeline/release-next/openshift/release/tektoncd-pipeline-nightly.yaml"
+STABLE_RELEASE_URL='https://raw.githubusercontent.com/openshift/tektoncd-pipeline/release-${version}/openshift/release/tektoncd-pipeline-${version}.yaml'
+
+function get_version {
+    echo get_get_version
+    local shift=${1} # 0 is latest, increase is the version before etc...
+    local version=$(curl -s https://api.github.com/repos/tektoncd/pipeline/releases | python -c "import sys, json;x=json.load(sys.stdin);print(x[${shift}]['tag_name'])")
+    echo version
+    echo  $version
+    echo $(eval echo ${STABLE_RELEASE_URL})
+}
+
+function tryurl {
+    curl -s -o /dev/null -f ${1} || return 1
+}
+
+function geturl() {
+    if [[ ${1} = "release-next" ]];then
+         if tryurl ${NIGHTLY_RELEASE};then
+             echo ${NIGHTLY_RELEASE}
+             return 0
+         fi
+         echo release next logic\n
+        for shifted in `seq 0 ${MAX_SHIFT}`;do
+            versionyaml=$(get_version ${shifted})
+            echo $versionyaml
```

Comment:
is that what you want to do ? echo versionyaml all the time here?

Reply:
and i think there is an issue here, this get_version doesnt install `release-next` (nightly) but a stable version.

The idea when I wrote the script on cli and catalog : 

- try install release nightly as tested on CI.
- if CI has failed  the release.yaml will not be here
- try to install the downstream stable release.
- we detect what is the latest release from upstream (as we don't have that information elsewhere)
- sometime downstream is lagging with upstream, so let's try a previous version of the release
- usually we should not be 1 version behind, if we are more than 1 we should really fail

Maybe the original script that I wrote is buggy (haven't checked) and doesn't respect that logic, if it doesn't then we would need to get it fixed.

Reply:
> is that what you want to do ? echo versionyaml all the time here?

no, that was unneccessary

Reply:
> and i think there is an issue here, this get_version doesnt install `release-next` (nightly) but a stable version.
> 
> The idea when I wrote the script is :
> 
> * try install release nightly as tested on CI.
> * if CI has failed  the release.yaml will not be here
> * try to install the downstream stable release.
> * we detect what is the latest release from upstream (as we don't have that information elsewhere)
> * sometime downstream is lagging with upstream, so let's try a previous version of the release
> * usually we should not be 1 version behind, if we are more than 1 we should really fail
> 
> Maybe the original script that I wrote is buggy (haven't checked) and doesn't respect that logic, if it doesn't then we would need to get it fixed.

The `function geturl()` has the `if block` for release-next (nightly)
    - it tries to check if the nightly url works
    - if no it will find a stable url (max shift 1)

the `else` block is for releases where we need a specific stable version.

Reply:
may i should write a separate script to handle payload fetching for nightly and for releases.

Reply:
current implementation fetches release-next for 
```update-to-head.sh```
and stable release for
```update-to-head.sh v0.9.2```

Reply:
what is your definition of release ? not tektoncd release but openshift-pipeline release?

Reply:
openshift-pipeline release(s)
available at 
```
STABLE_RELEASE_URL='https://raw.githubusercontent.com/openshift/tektoncd-pipeline/release-${version}/openshift/release/tektoncd-pipeline-${version}.yaml'
```

as this repo has no upstream every reference is from openshift org